src/dataframe/dataframe.jl

```diff
@@ -1317,3 +1317,79 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+function _expandhelper(col, vect, stride, offset)
+    for i in offset:stride:length(vect)
+        push!(col, vect[i])
+    end
+end
+
+function expand(df::AbstractDataFrame, indexcols)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    if !(indexcols isa Symbol || allunique(indexcols))
+        throw(ArgumentError("Elements of $indexcols must be unique"))
+    end
+    sum(nonunique(df, indexcols))>0 && @warn "duplicate rows in input; expand will only return unique combinations"
+    colind = index(df)[indexcols]
+    dummydf = similar(select(df, colind, copycols=false), 0)
+
+    # Create a vect of vectors of unique values in each column
+    uniqueVals = []
+    for col in colind
+        # levels drops missing, handle the case where missing values are present
+        if any(ismissing, df[!, col])
+            tempcol = vcat(levels(df[col]), missing)
+        else
+            tempcol = levels(df[col])
+        end
+        push!(uniqueVals, tempcol)
+    end
+
+    # Get a long vector of every possible combination
+    collected = collect(Iterators.flatten(Iterators.product(uniqueVals...)))
```

Comment:
Now I have run your code. It will be much faster not to `Iterators.flatten` this iterator but rather to something like:

```
p = Iterators.product(uniqueVals...)
for i in 1:length(indexcols)
    _expandhelper(dummydf[!, i], p, i)
end
```

and define `_expandhelper` as:

```
function _expandhelper(col, prod, idx)
    for v in prod
        push!(col, v[i])
    end
end
```
the reason is that we do not create intermediate container with abstract eltype.

Reply:
Just tested this and this indeed ended up being faster. Was the intermediate container that you were referring to the flattened Iterator?